flake: fix `lib.trivial.version` when used from a flake

Ma27 opened the pull request:
## Description of changes
A lot of fetchers from Nix's own `libfetchers` also provide the information that `lib.trivial` aims to expose with `version`/`versionSuffix`/`revision`. In fact you don't even need a `nixpkgs` channel to get a proper version suffix because of that!

Unfortunately this isn't used currently. When using the nixpkgs flake, but not `nixpkgs.lib.nixosSystem` to build a NixOS configuration, the version will always be `YY.MMpre-git`. One example is e.g. `colmena` which evaluates configurations via
`import (npkgs.path + "/nixos/lib/eval-config.nix")`.

This patch ensures that the version suffix (i.e. the normalized last modified date + git revision) is correctly exposed in `lib.trivial`. Additionally, the change is injected into the following locations:

* `lib`: with that, something like

      $ nix eval nixpkgs#lib.trivial.version
      23.05.20230921.cf8bf79

  is working fine (i.e. rather than `23.05pre-git`).

* `legacyPackages` to make sure that e.g. `legacyPackages.<system>.nixos` has correct version info. This also applies to everything else using `pkgs.lib.trivial` for that purpose.

* `overlays.default` which can be applied to a `nixpkgs` and changes the previous `pkgs.lib` from said `nixpkgs` to also contain the correct `version`/`revision`/etc..

  This is useful for people using `nixpkgs` as flake input, but importing it manually with

      import inputs.nixpkgs { }

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Ma27 committed:
flake: fix `lib.trivial.version` when used from a flake

A lot of fetchers from Nix's own `libfetchers` also provide the
information that `lib.trivial` aims to expose with
`version`/`versionSuffix`/`revision`. In fact you don't even need a
`nixpkgs` channel to get a proper version suffix because of that!

Unfortunately this isn't used currently. When using the
nixpkgs flake, but not `nixpkgs.lib.nixosSystem` to build a NixOS
configuration, the version will always be `YY.MMpre-git`. One example is
e.g. `colmena` which evaluates configurations via
`import (npkgs.path + "/nixos/lib/eval-config.nix")`.

This patch ensures that the version suffix (i.e. the normalized last
modified date + git revision) is correctly exposed in `lib.trivial`.
Additionally, the change is injected into the following locations:

* `lib`: with that, something like

      $ nix eval nixpkgs#lib.trivial.version
      23.05.20230921.cf8bf79

  is working fine (i.e. rather than `23.05pre-git`).

* `legacyPackages` to make sure that e.g. `legacyPackages.<system>.nixos`
  has correct version info. This also applies to everything else using
  `pkgs.lib.trivial` for that purpose.

* `overlays.default` which can be applied to a `nixpkgs` and changes the
  previous `pkgs.lib` from said `nixpkgs` to also contain the correct
  `version`/`revision`/etc..

  This is useful for people using `nixpkgs` as flake input, but
  importing it manually with

      import inputs.nixpkgs { }

Co-authored-by: Linus Heckemann <git@sphalerite.org>

roberth commented on flake.nix:
```diff
@@ -9,24 +9,31 @@
+          version = finalLib.trivial.release + finalLib.trivial.versionSuffix;
+          revisionWithDefault = default: self.rev or default;
+        };
+      };
```
This logic should also apply to the `lib/flake.nix` flake, or ideally be moved.

Ma27 replied:
> lib/flake.nix flake, or ideally be moved.

TIL about `lib/flake.nix`. Will move the implementation there.

Ma27 replied:
Hmm... I'm not sure if we want to have a flake-input for nixpkgs (which would be necessary when moving the code to lib's flake.nix).

Other options I see:
* just duplicate it and add a warning on top that the code needs to be kept in sync with lib's flake
* move the lib overlay somehow into lib itself. Doesn't seem too nice to me: until now flakes are an entry-point only to nixpkgs and lib, with that change we'd assume flakes in lib itself which seems wrong to be since lib should be usable in any Nix evaluation, right? Also, wouldn't that give wrong (or misleading) results when evaluating e.g. `self.shortRev` from `lib/flake.nix`?

roberth replied:
> I'm not sure if we want to have a flake-input for nixpkgs

I don't see a problem with that; only an inconvenience, which is lock file pollution, which is solvable and should be solved on the Nix side.
- https://github.com/NixOS/nix/issues/7730

Oh and subflakes are still broken. That's an actual problem. Never mind.



> * move the lib overlay somehow into lib itself.

We could have a separate "internal" file in `lib/` that takes care of implementing the version setting, without polluting the `lib` attrset. Both flakes could use that file.



> * Also, wouldn't that give wrong (or misleading) results when evaluating e.g. `self.shortRev` from `lib/flake.nix`?

The whole business about versions is error prone if your expectations of it are too high. Not sure how to manage those expectations.

roberth reviewed (changes requested):
Generally good idea, but not quite complete.

Ma27 commented on flake.nix:
```diff
@@ -9,24 +9,31 @@
     {
+      overlays.default = final: prev: {
+        lib = prev.lib.extend libVersionInfoOverlay;
+      };
```
Fixed the naming of the overlay.

Won't resolve though until we're aligned on how to deal with `import inputs.nixpkgs`.

Ma27 replied:
Is the naming OK and the fact that you'd have to apply this manually when doing `import inputs.nixpkgs` @roberth? Also, anything else to discuss / fix? :)

roberth replied:
I think we should discourage `import inputs.nixpkgs`, because it just can't work well. It's just not something that can ever work as well as the actual flake outputs, unless flakes allows `outPath` to be a `__functor` function. I really don't know what to think of that... It should probably be an explicit feature of flakes and not something that allows arbitrary stuff to be put in `outPath`. And even then it's kind of bad because of `import "${inputs.nixpkgs}"`.

Ignoring all that ugliness for now, I think the `nixpkgs` should provide a proper `lib`-based entrypoint into Nixpkgs, just like we have `lib.nixosSystem`. That way it can actually work well, and not be a strange experience. (`import` a flake, really?)

Some possible names
- `lib.nixpkgs`: not a fan because Nixpkgs *should* be more than just the function that returns "pkgs"
- `lib.callNixpkgs`: arbitrary enough that we don't paint ourselves into a corner just yet

We should probably get the arguments right this time, along the lines of
`{ hostPlatform, buildPlatform ? hostPlatform, config , overlays }:` instead of the unholy mess that we have in `default.nix`. Maybe `default.nix` should stay bad, as a motivator for people to use the better entrypoint?

(As mentioned in the review, I think we should solve all these problems in a new PR after merging this one, and not have `overlays` in the outputs for this)

Ma27 committed:
flake: also provide proper version info for lib's flake

This effectively means that

    nixpkgs$ nix eval ./lib#lib.trivial.version
    "23.11.20231020.ee0d6b5"

now gives meaningful results as well.

See https://github.com/NixOS/nixpkgs/pull/257100#discussion_r1352075369
for the discussion around this.

roberth commented on lib/flake.nix:
```diff
@@ -1,5 +1,10 @@
-  outputs = { self }: { lib = import ./.; };
+  outputs = { self }:
+    let
+      lib' = import ./.;
+    in {
+      lib = lib'.extend (import ./__flake-version-info.nix self);
```
`'` is often used for the second state rather than the first, particularly in math.

```suggestion
      lib0 = import ./.;
    in {
      lib = lib0.extend (import ./__flake-version-info.nix self);
```

```suggestion
      libBase = import ./.;
    in {
      lib = libBase.extend (import ./__flake-version-info.nix self);
```

roberth commented on lib/__flake-version-info.nix:
```diff
@@ -0,0 +1,17 @@
+# lib overlay to be used by the nixpkgs & nixpkgs/lib flakes
+# to provide meaningful values for `lib.trivial.version` et al..
+#
+# Internal and subject to change, don't use this anywhere else!
```
```suggestion
# Internal and subject to change, don't use this anywhere else!
# Instead, consider using a public interface, such as this flake here in this directory, `lib/`, or
# use the nixpkgs flake, which applies this logic for you in its `lib` output attribute.
```

roberth commented on lib/__flake-version-info.nix:
```diff
@@ -0,0 +1,17 @@
+# lib overlay to be used by the nixpkgs & nixpkgs/lib flakes
```
```suggestion
# This function produces a lib overlay to be used by the nixpkgs & nixpkgs/lib flakes
```

nit: this file has a _function to_ an overlay - not quite an overlay yet.

roberth reviewed (approved):
Only nit picks. Would like another reviewer to have a look.
Maybe @infinisil or @lheckemann have opinions?

infinisil commented on lib/__flake-version-info.nix:
Files are internal by default, we shouldn't need underscores to indicate that.

Ma27 replied:
Not sure I agree given that each flake input is also a path, but I'm fine with changing that.

infinisil commented on lib/__flake-version-info.nix:
```diff
@@ -0,0 +1,17 @@
+
+{
+  trivial = prevLib.trivial // {
+    versionSuffix =
+      ".${finalLib.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}.${self.shortRev or "dirty"}";
+    version = finalLib.trivial.release + finalLib.trivial.versionSuffix;
+    revisionWithDefault = default: self.rev or default;
```
This is partially duplicated code and is not very re-usable. Instead we can add these to proper `lib.trivial`:
- `lib.trivial.lastModifiedDate` (no `lastModified` since now Flakes always provide `lastModifiedDate`), defaulting to `null`.
- `lib.trivial.revision` (maybe also a `shortRev`, but not necessary since it can easily be computed), defaulting to what `revisionWithDefault` currently defaults to.

Then there's no need to have any implementation here, because it can be moved to the original definitions for `versionSuffix` and `revisionWithDefault`. You will have to remove the `rec` from `lib/trivial.nix` though in order for `version` to be updated correctly.

Then this here can just be

```nix
{
  lastModifiedDate = self.lastModifiedDate;
  revision = self.rev;
}
```

This also allows us to test and document this functionality in `lib/tests`, which I really think should be done.

roberth replied:
> `trivial`

I don't think this is insignificant or self-evident.

> re-usable

Not sure if that should be a goal here.

> test and document this functionality

Good point, but `lib/tests` can import a file like this to achieve that.

Ma27 replied:
> like these to proper lib.trivial:

To make sure I'm not misunderstanding you: you're talking about `lib/trivial.nix` itself, correct?

> no lastModified since now Flakes always provide lastModifiedDate

Good point. Was there for historical reasons only I guess.

> You will have to remove the rec from lib/trivial.nix though in order for version to be updated correctly.

That can be done independently of the other changes you suggested here IMHO.

That said, the downside is that you'll have to touch the entire code affecting `.version-suffix` because that's where the `{lastmodified}.{shortrev}` thing actually comes from in the non-flake case.

> but not necessary since it can easily be computed

Not necessarily because git IIRC determines a length that makes each short rev unique (I think hardcoding a short-rev to e.g. 8 may cause a few collisions already).

infinisil replied:
> you're talking about `lib/trivial.nix` itself, correct?

Yes.



> That said, the downside is that you'll have to touch the entire code affecting `.version-suffix` because that's where the `{lastmodified}.{shortrev}` thing actually comes from in the non-flake case.

No I'm thinking to change the definition of `revisionWithDefault` https://github.com/NixOS/nixpkgs/blob/0a1051cba784efc392695736399c17e51e687e44/lib/trivial.nix#L212-L221 to have an extra `if` condition to use `lib.trivial.revision` if it's not `null`. But I'm not insisting on that, I think it's okay the way this PR is now.



> Not necessarily because git IIRC determines a length that makes each short rev unique (I think hardcoding a short-rev to e.g. 8 may cause a few collisions already).

Ah true, I guess the short revs would still be prone to collisions over longer time though. I'd prefer a sufficiently long constant short rev length that is reasonably collision resistant for the future.

Ma27 replied:
>  to have an extra if condition to use lib.trivial.revision if it's not null. But I'm not insisting on that, I think it's okay the way this PR is now.

Hmm I don't think I like that: having an attribute that only works magically when using flakes and that provides a revision despite all the other attributes for that seems unintuitive to me at best.

IMHO redesigning this version code here is another topic and the what we have here is quite an improvement already, so I'd actually prefer the current version for now.

infinisil commented on lib/__flake-version-info.nix:
```diff
@@ -0,0 +1,17 @@
+{
+  trivial = prevLib.trivial // {
+    versionSuffix =
+      ".${finalLib.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}.${self.shortRev or "dirty"}";
```
This feels very arbitrary, but I guess it was done that way before. This should be documented as using an unspecified format.

roberth replied:
Hyrum's law:

> With a sufficient number of users of an API,
> it does not matter what you promise in the contract:
> all observable behaviors of your system
> will be depended on by somebody.

I figure for something like this, we might as well promise not to change it, and even document the exact format.

Ma27 replied:
> This feels very arbitrary, but I guess it was done that way before

@infinisil Correct, this is precisely the format everybody had when using `nixpkgs.lib.nixosSystem`.

It's not entirely unspecified or arbitrary though. This is the format you also get when using a channel tarball to build your system:

```
$ pwd                               
/nix/var/nix/profiles/per-user/root/channels/nixos
$ cat .version-suffix 
.20231011.97d028b
$ nix-instantiate --eval -A lib.trivial.versionSuffix 
".20231011.97d028b"
```

Only when importing the nixpkgs flake as path rather than as git input, you'll get something else, i.e. `YYYYMMDD.dirty`. I'd argue that `YYYYMMDD.dirty` is still better than `pre-git` though since you get a little more information at least (what we can't implement with lib.trivial and channel tarballs I think). However we can switch back to `pre-git` here if you prefer that (while I prefer the former option, I don't have a strong opinion though because you should only encounter that when testing something with a local checkout).

I'll update the documentation accordingly as soon as we've agreed on everything else.

> we might as well promise not to change it

@roberth fwiw Hyrum's law should already apply already since the format was/is used by `nixpkgs.lib.nixosSystem`.

infinisil replied:
I think it's fine to provide as much info as possible, let's treat the version suffix as an arbitrary string without guarantees.

infinisil replied:
Where are you thinking of documenting this though? Because it wouldn't really fit into the docs for `lib.trivial.versionSuffix`, since it would be specific to Flakes. I guess we could close an eye for that, but I feel like anything documented in `lib` should apply generally.

Ma27 replied:
> Where are you thinking of documenting this though?

Didn't bother to find a suitable place yet because I wanted the other stuff to get resolved first :upside_down_face: 

> Because it wouldn't really fit into the docs for lib.trivial.versionSuffix, since it would be specific to Flakes

Oh so we have some basic documentation about that already? I'm afraid I fail to find it right now, but may I ask where?

> since it would be specific to Flakes

It's even just the "edge" case of a local path override where the suffix looks different, so I'd argue that it would be OKish.

> but I feel like anything documented in lib should apply generally.

I mean, we effectively have two ways of using lib (this change also applies to the `nixpkgs/lib` flake itself), so it seems OK to me to mention a slightly different behavior when using one of them.

infinisil replied:
> Oh so we have some basic documentation about that already? I'm afraid I fail to find it right now, but may I ask where?

[We do indeed](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.trivial.versionSuffix) :slightly_smiling_face: ([source](https://github.com/NixOS/nixpkgs/blob/8ad872b2f923793abc8822cce462fbad03dec237/lib/trivial.nix#L200))

lheckemann replied:
That already doesn't specify what the suffix contains, right? And this change doesn't remove that documentation, so I think it's fine as is?

Ma27 replied:
Good point. This string is for informational purposes only (i.e. when I want to check what's deployed on my system, then this is helpful), so perhaps it's not a good idea to start documenting a rather arbitrary format in there now. Hence, resolving this thread now.

infinisil reviewed (changes requested):
The idea is okay, but I think it should be implemented differently.

Ma27 committed:
lib/trivial: drop `rec` in favor of `lib` fixpoint

That way each expression uses the final version of other lib.trivial
declarations.

For instance, when replacing `versionSuffix` with the string `"fnord"`
in a lib overlay, `trivial.version` uses `"fnord"` as suffix now rather
than `pre-git`.

Ma27 committed:
flake/version overlay: review fixes

* Improves the comments of `lib/flake-version-info.nix` and drops the
  `__`-prefix from the filename.
* `lib'` -> `lib0` in `nixpkgs/lib`.
* Drop the declaration of `trivial.version` in the overlay because this
  declaration already uses the final expressions of `versionSuffix` and
  `release` now.
* No need to fall back to `self.lastModified` anymore, this was a
  workaround for pre2.4 Nix.

Co-authored-by: Robert Hensing <robert@roberthensing.nl>
Co-authored-by: Silvan Mosberger <contact@infinisil.com>

Ma27 commented:
Implemented a few of the suggestions, namely

* comment improvements from @roberth
* dropped the `rec` in `lib/trivial.nix` to spare the `version =` expression.
* dropped the `__`-prefix from the lib overlay's filename.
* dropped the fallback to `lastModified` when generating the version suffix.

Re introducing `lib.trivial.lastModifiedDate`/`lib.trivial.revision` I left a comment above because that'd mean a more significant change to nixpkgs internals that I'd like to discuss first.

As mentioned above, I'll update the documentation about the corner cases of flake's versionSuffix when everything is done (in the default case it behaves the same as when using channel tarballs).

infinisil commented on flake.nix:
```diff
@@ -53,7 +52,10 @@
       # which keeps `nix flake show` on Nixpkgs reasonably fast, though less
       # information rich.
-      legacyPackages = forAllSystems (system: import ./. { inherit system; });
+      legacyPackages = forAllSystems (system: import ./. {
+        inherit system;
+        overlays = [ self.overlays.setLibVersionInfo ];
+      });
```
I believe this is not entirely backwards compatible, because if you use `--impure`, `overlays` defaults to stuff from `~/.config/nixpkgs`, whereas passing the argument directly will prevent that from happening. So instead we should just extend it with the overlay:
```suggestion
      legacyPackages = forAllSystems (system:
        (import ./. { inherit system; }).extend self.overlays.setLibVersionInfo
      );
```

Ma27 replied:
Nice catch, thanks!

infinisil reviewed (approved):
Just a minor detail, this looks good to me otherwise!

Ma27 committed:
flake: be backwards-compatible for `--impure`

We cannot pass `overlays = ...` to `nixpkgs` directly because by default
overlays from `~/.config/nixpkgs` are loaded in there. This doesn't
happen by default, but when using `--impure`.

Explicitly specifying that ignores these overlays. By using `pkgs.extend`
the old behavior can be kept and the new overlay can be applied.

Co-authored-by: Silvan Mosberger <contact@infinisil.com>

Ma27 commented:
Fixed @infinisil's remark, rebased onto master and also fixed an evaluation error in `trivial.nix` (I dropped the `rec` in there and a newly added function (`mirrorFunctionArgs`) broke because of that).

Any final remarks or can we merge @roberth @lheckemann @infinisil ?

roberth commented on lib/trivial.nix:
```diff
@@ -446,7 +446,7 @@ rec {
   */
   isFunction = f: builtins.isFunction f ||
-    (f ? __functor && isFunction (f.__functor f));
+    (f ? __functor && lib.trivial.isFunction (f.__functor f));
```
```suggestion
    (f ? __functor && isFunction (f.__functor f));
```

Performance regression. Use `let inherit (lib) isFunction;` near start of file.

roberth replied:
Please do the same for the rest - at least for everything that doesn't perform I/O like `warn`.

Ma27 replied:
Done.

Btw I'm curious, is "dereferencing" an attribute-set really that more expensive? (not _that_ deep into the evaluator I must admit :sweat_smile: )

roberth replied:
No, it'd only be measurable if it's in a hot path, but functions like `isFunction` probably are. Not worth risking.
Also less noise.

roberth commented on flake.nix:
```diff
@@ -9,24 +9,23 @@
       forAllSystems = lib.genAttrs lib.systems.flakeExposed;
     in
     {
+      overlays.setLibVersionInfo = final: prev: {
+        lib = prev.lib.extend libVersionInfoOverlay;
+      };
```
```suggestion
```

This is not a Nixpkgs overlay, but a `lib` overlay. It seems unnecessary, and confusing.

We can discuss it more, or first merge without this, because it is not essential.

Ma27 replied:
Fine for me, done.

roberth reviewed (commented):
Can merge after implementing the suggestions.

Scoping out [nixpkgs entrypoint problems](https://github.com/NixOS/nixpkgs/pull/257100#discussion_r1421684826) for now.

Ma27 committed:
flake: drop libVersionInfoOverlay

It's not really a pkgs overlay, but a lib overlay.
